packages/protocol/contracts/governance/DoubleSigningSlasher.sol

```diff
@@ -0,0 +1,179 @@
+    address signer,
+    uint256 index,
+    bytes memory blockA,
+    bytes memory blockB
+  ) public view returns (uint256) {
+    require(hashHeader(blockA) != hashHeader(blockB), "Block hashes have to be different");
+    uint256 blockNumber = getBlockNumberFromHeader(blockA);
+    require(
+      blockNumber == getBlockNumberFromHeader(blockB),
+      "Block headers are from different height"
+    );
+    require(index < numberValidatorsInSet(blockNumber), "Bad validator index");
+    require(
+      signer == validatorSignerAddressFromSet(index, blockNumber),
+      "Wasn't a signer with given index"
+    );
+    uint256 mapA = uint256(getVerifiedSealBitmapFromHeader(blockA));
+    uint256 mapB = uint256(getVerifiedSealBitmapFromHeader(blockB));
+    require(mapA & (1 << index) != 0, "Didn't sign first block");
+    require(mapB & (1 << index) != 0, "Didn't sign second block");
+    require(
+      countSetBits(mapA) >= minQuorumSize(blockNumber),
+      "Not enough signers in the first block"
+    );
+    require(
+      countSetBits(mapB) >= minQuorumSize(blockNumber),
+      "Not enough signers in the second block"
+    );
+    return blockNumber;
+  }
+
+  /**
+   * @notice Requires that `eval` returns true and that this evidence has not
+   * already been used to slash `signer`.
+   * If so, fetches the `account` associated with `signer` and the group that
+   * `signer` was a member of during the corresponding epoch.
+   * Then, calls `LockedGold.slash` on both the validator and group accounts.
+   * Calls `Validators.removeSlashedMember` to remove the validator from its
+   * current group if it is a member of one.
+   * Finally, stores that hash(signer, blockNumber) has been slashed.
+   * @param signer The signer to be slashed.
+   * @param index Validator index at the block.
+   * @param headerA First double signed block.
+   * @param headerB Second double signed block.
+   * @param groupMembershipHistoryIndex Group membership index from where the group should be found.
+   * @param validatorElectionLessers Lesser pointers for validator slashing.
+   * @param validatorElectionGreaters Greater pointers for validator slashing.
+   * @param validatorElectionIndices Vote indices for validator slashing.
+   * @param groupElectionLessers Lesser pointers for group slashing.
+   * @param groupElectionGreaters Greater pointers for group slashing.
+   * @param groupElectionIndices Vote indices for group slashing.
+   */
+  function slash(
+    address signer,
+    uint256 index,
+    bytes memory headerA,
+    bytes memory headerB,
+    uint256 groupMembershipHistoryIndex,
+    address[] memory validatorElectionLessers,
+    address[] memory validatorElectionGreaters,
+    uint256[] memory validatorElectionIndices,
+    address[] memory groupElectionLessers,
+    address[] memory groupElectionGreaters,
+    uint256[] memory groupElectionIndices
+  ) public {
+    require(!isSlashed[signer][hashHeader(headerA)], "Already slashed");
```

Comment:
is `hashHEader` and expensive operation?
we are using it 3 times for each block. 2 here, and one in `checkForDoubleSigning`

Reply:
Changed it so that it now only does it twice...